cmd/opampsupervisor/supervisor/packages.go

```diff
@@ -0,0 +1,474 @@
+	}
+
+	checkOpts, err := createCosignCheckOpts(signatureOpts)
+	if err != nil {
+		return nil, fmt.Errorf("create signature verification options: %w", err)
+	}
+
+	return &packageManager{
+		packageState:    state,
+		topLevelHash:    agentHash,
+		topLevelVersion: agentVersion,
+		storageDir:      storageDir,
+		agentPath:       agentPath,
+		checkOpts:       checkOpts,
+		am:              am,
+	}, nil
+}
+
+func (p packageManager) AllPackagesHash() ([]byte, error) {
+	return p.packageState.AllPackagesHash, nil
+}
+
+func (p packageManager) SetAllPackagesHash(hash []byte) error {
+	p.packageState.AllPackagesHash = hash
+	return savePackageState(p.packagesStatusPath(), p.packageState)
+}
+
+func (packageManager) Packages() ([]string, error) {
+	return []string{agentPackageKey}, nil
+}
+
+func (p packageManager) PackageState(packageName string) (state types.PackageState, err error) {
+	if packageName == agentPackageKey {
+		return types.PackageState{
+			Exists:  true,
+			Type:    protobufs.PackageType_PackageType_TopLevel,
+			Hash:    p.topLevelHash,
+			Version: p.topLevelVersion,
+		}, nil
+	}
+
+	return types.PackageState{
+		Exists: false,
+	}, nil
+}
+
+func (p *packageManager) SetPackageState(packageName string, state types.PackageState) error {
+	if packageName != agentPackageKey {
+		return fmt.Errorf("package %q does not exist", packageName)
+	}
+
+	if !state.Exists {
+		return fmt.Errorf("agent package must be marked as existing")
+	}
+
+	if state.Type != protobufs.PackageType_PackageType_TopLevel {
+		return fmt.Errorf("agent package must be marked as top level")
+	}
+
+	p.topLevelHash = state.Hash
+	p.topLevelVersion = state.Version
+
+	return nil
+}
+
+func (packageManager) CreatePackage(packageName string, _ protobufs.PackageType) error {
+	if packageName != agentPackageKey {
+		return fmt.Errorf("only agent package is supported")
+	}
+
+	return fmt.Errorf("agent package already exists")
+}
+
+func (p *packageManager) FileContentHash(packageName string) ([]byte, error) {
+	if packageName != agentPackageKey {
+		return nil, nil
+	}
+
+	return p.topLevelHash, nil
+}
+
+func (p *packageManager) UpdateContent(ctx context.Context, packageName string, data io.Reader, contentHash, signature []byte) error {
+	if packageName != agentPackageKey {
+		return fmt.Errorf("package does not exist")
+	}
+
+	by, err := io.ReadAll(data)
+	if err != nil {
+		return fmt.Errorf("read package bytes: %w", err)
+	}
+
+	err = verifyPackageIntegrity(by, contentHash)
+	if err != nil {
+		return fmt.Errorf("could not verify package integrity: %w", err)
+	}
+
+	b64Cert, b64Signature, err := parsePackageSignature(signature)
+	if err != nil {
+		return fmt.Errorf("could not parse package signature: %w", err)
+	}
+
+	err = verifyPackageSignature(ctx, p.checkOpts, by, b64Cert, b64Signature)
+	if err != nil {
+		return fmt.Errorf("could not verify package signature: %w", err)
+	}
+
+	// overwrite agent process
+	startAgent, err := p.am.stopAgentProcess(ctx)
```

Comment:
My thought was that we need to stop the agent process to write the binary in the first place.

We could write the binary somewhere temporarily, and send a message to the agent goroutine to do the replacement, which could be cleaner (better separation of concerns). The package manager would still need to have some handle to signal that, but it wouldn't necessarily be tied to the collector.

Reply:
Would we be able to accomplish this with synchronous, one-way communication? I was thinking maybe we could do something vaguely like this in the Supervisor struct:

```golang
oldVersion := 0
newVersion := 1

s.commander.stopAgent()
s.packageManager.SwapInVersion(newVersion)

if err := s.commander.startAgent(); err != nil {
	s.packageManager.SwapInVersion(oldVersion)
	s.commander.startAgent()
	s.reportFailedInstallation()
	return
}

s.packageManager.DeleteVersion(oldVersion)
s.reportSuccessfulInstallation()
```

Reply:
Yeah, I think we could do something like that. I think this also sets us up nicely for restoring the previous version if the new one fails. I'll look into making those changes.

Reply:
Alright, so I'm looking a bit deeper and the reason I'm a little iffy on this is because it seems like there will be some odd consequences with doing this outside of the package syncer.

Essentially, the package syncer will run it's logic, then it will report the status off every package at the end. So it's possible that the package downloads and verifies fine, so the package syncer reports the package as installed - but then there is an error when swapping in the version, which causes an error.

I would prefer if we didn't report the package is installed before it's actually successfully installed, basically. Which means we'd have to do it in the package manager during the `UpdateContent` call, as far as I can tell. This would require 2 way communication, one way from (packagemanager) -> (agent goroutine) to do the swap, then one from (agent goroutine) -> (package manager) to indicate success/failure.